napari/utils/progress.py

```diff
@@ -0,0 +1,138 @@
+import inspect
+from typing import Iterable, Optional
+
+from tqdm import tqdm
+
+_tqdm_kwargs = {
+    p.name
+    for p in inspect.signature(tqdm.__init__).parameters.values()
+    if p.kind is not inspect.Parameter.VAR_KEYWORD and p.name != "self"
+}
+
+
+class progress(tqdm):
+    """This class inherits from tqdm and provides an interface for
+    progress bars in the napari viewer. Progress bars can be created
+    directly by wrapping an iterable or by providing a total number
+    of expected updates.
+
+    See tqdm.tqdm API for valid args and kwargs:
+    https://tqdm.github.io/docs/tqdm/
+
+    Also, any keyword arguments to the :class:`ProgressBar` `QWidget`
+    are also accepted and will be passed to the ``ProgressBar``.
+
+    Examples
+    --------
+
+    >>> def long_running(steps=10, delay=0.1):
+    ...     for i in progress(range(steps)):
+    ...         sleep(delay)
+
+    it can also be used as a context manager:
+
+    >>> def long_running(steps=10, repeats=4, delay=0.1):
+    ...     with progress(range(steps)) as pbr:
+    ...         for i in pbr:
+    ...             sleep(delay)
+
+    or equivalently, using the `progrange` shorthand
+    ...     with progrange(steps) as pbr:
+    ...         for i in pbr:
+    ...             sleep(delay)
+
+    For manual updates:
+
+    >>> def manual_updates(total):
+    ...     pbr = progress(total=total)
+    ...     sleep(10)
+    ...     pbr.set_description("Step 1 Complete")
+    ...     pbr.update(1)
+    ...     # must call pbr.close() when using outside for loop
+    ...     # or context manager
+    ...     pbr.close()
+
+    """
+
+    def __init__(
+        self,
+        iterable: Optional[Iterable] = None,
+        desc: Optional[str] = None,
+        total: Optional[int] = None,
+        *args,
+        **kwargs,
+    ) -> None:
+        kwargs = kwargs.copy()
+        pbar_kwargs = {k: kwargs.pop(k) for k in set(kwargs) - _tqdm_kwargs}
+
+        # get progress bar added to viewer
+        try:
+            from .._qt.widgets.qt_progress_bar import get_pbar  # noqa
```

Comment:
This is definitely not a great smell... I see a lot of value in getting the progress bar implementation out in the next release, but I'd like to make an issue to remind us to clean this up... I think ideally we separate this as an *emitting* tqdm progress bar, and create a qt progress bar that subscribes to those events. Then users would do `from napari.qt import progress` and get the widget.

Or is that too heavyweight...? I dunno. But it's definitely weird to have qt things outside of the `_qt/` and `qt/` sandboxes...

Reply:
yeahh I wasn't too sure about this either. The whole point of this class is to provide a GUI-viewable progress bar. From that perspective I agree that maybe it should live in qt alongside its widget companions. I initially moved the import into the `__init__` to support failing semi-gracefully back to just terminal `tqdm` if there's no viewer. 

> I think ideally we separate this as an emitting tqdm progress bar, and create a qt progress bar that subscribes to those events. Then users would do from napari.qt import progress and get the widget.

Not 100% sure what you mean here, sorry.

Reply:
Ok yeah, I think this should be moved inside our `qt` space and imported from there. I would do that now before merge. Good catch @jni

Reply:
Will move that soon (I've got class for two hours now). Should we still keep the `import` in the `__init__` just in case people do want the option to fall back to `tqdm` in the terminal if they've got long running scripts that may or may not be run with a viewer?

Reply:
hmm i see the trickiness. I'd probably just move it all for now, we can always expose more later - but harder to take something away later, but i'll defer to @jni and @tlambert03 if they have strong feelings here

Reply:
> This is definitely not a great smell... I see a lot of value in getting the progress bar implementation out in the next release, but I'd like to make an issue to remind us to clean this up... I think ideally we separate this as an *emitting* tqdm progress bar, and create a qt progress bar that subscribes to those events. Then users would do `from napari.qt import progress` and get the widget.
> 
> 
> 
> Or is that too heavyweight...? I dunno. But it's definitely weird to have qt things outside of the `_qt/` and `qt/` sandboxes...

That's what viewer.window is too!  The general concept of things that use qt if available but otherwise have functionality needn't be bad smell/design